src/hotspot/share/code/codeCache.cpp

```diff
@@ -840,7 +956,50 @@ void CodeCache::purge_exception_caches() {
   _exception_cache_purge_list = NULL;
 }
 
+// Register an is_unloading nmethod to be flushed after unlinking
+void CodeCache::register_unlinked(nmethod* nm) {
+  assert(nm->unlinked_next() == NULL, "Only register for unloading once");
+  for (;;) {
+    // Only need acquire when reading the head, when the next
+    // pointer is walked, which it is not here.
+    nmethod* head = Atomic::load(&_unlinked_head);
+    nmethod* next = head != NULL ? head : nm; // Self looped means end of list
+    nm->set_unlinked_next(next);
+    if (Atomic::cmpxchg(&_unlinked_head, head, nm) == head) {
+      break;
+    }
+  }
+}
+
+// Flush all the nmethods the GC unlinked
+void CodeCache::flush_unlinked_nmethods() {
+  nmethod* nm = _unlinked_head;
+  _unlinked_head = NULL;
+  size_t freed_memory = 0;
+  while (nm != NULL) {
+    nmethod* next = nm->unlinked_next();
+    freed_memory += nm->total_size();
+    nm->flush();
+    if (next == nm) {
+      // Self looped means end of list
```

Comment:
Why not nullptr for the end of the list?

Reply:
Because inline caches. When the end is self looped instead of null, I can check if an nmethod is on the list or not, which is used to ensure that unlinking of an nmethod during class unloading happens only once. Why would it ever happen twice? Because the task that unlinks nmethod might have to restart due to inline cache cleaning failing due to running out of ICStubs. And when the code cache is walked the second time, I don't want to call unlink again if we already did that the first traversal. That's why the end of list is self looped instead of null.